src/main/groovy/com/google/protobuf/gradle/ProtobufModelPlugin.groovy

```diff
@@ -0,0 +1,251 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.protobuf.gradle
+
+import com.google.common.collect.Maps
+import org.gradle.api.*
+import org.gradle.api.file.FileCollection
+import org.gradle.api.file.SourceDirectorySet
+import org.gradle.api.internal.file.DefaultSourceDirectorySet
+import org.gradle.api.internal.file.FileResolver
+import org.gradle.api.plugins.ExtensionContainer
+import org.gradle.api.tasks.TaskContainer
+import org.gradle.api.tasks.util.PatternSet
+import org.gradle.internal.reflect.Instantiator
+import org.gradle.internal.service.ServiceRegistry
+import org.gradle.language.base.LanguageSourceSet
+import org.gradle.language.base.internal.SourceTransformTaskConfig
+import org.gradle.language.base.internal.registry.LanguageTransform
+import org.gradle.language.base.internal.registry.LanguageTransformContainer
+import org.gradle.language.base.plugins.ComponentModelBasePlugin
+import org.gradle.language.base.sources.BaseLanguageSourceSet
+import org.gradle.language.cpp.CppSourceSet
+import org.gradle.language.cpp.internal.DefaultCppSourceSet
+import org.gradle.language.cpp.tasks.CppCompile
+import org.gradle.model.*
+import org.gradle.nativeplatform.NativeBinarySpec
+import org.gradle.nativeplatform.NativeComponentSpec
+import org.gradle.nativeplatform.NativeDependencySet
+import org.gradle.platform.base.BinarySpec
+import org.gradle.platform.base.LanguageType
+import org.gradle.platform.base.LanguageTypeBuilder
+import org.gradle.platform.base.TransformationFileType
+import org.gradle.util.CollectionUtils
+
+import javax.inject.Inject
+import java.util.concurrent.Callable
+
+class ProtobufModelPlugin implements Plugin<Project> {
+  private final FileResolver fileResolver
+
+  @Inject
+  public ProtobufModelPlugin(FileResolver fileResolver) {
+    this.fileResolver = fileResolver;
+  }
+
+  @Override
+  void apply(Project project) {
+    def gv = project.gradle.gradleVersion =~ "(\\d*)\\.(\\d*).*"
+    if (!gv || !gv.matches() || gv.group(1).toInteger() < 2 || gv.group(2).toInteger() < 7) {
+      println("You are using Gradle ${project.gradle.gradleVersion}: This version of the protobuf" +
+          " plugin for c++ requires minimum Gradle version 2.7")
+    }
+
+    // Provides the osdetector extension
+    project.pluginManager.apply('osdetector')
+
+    project.extensions.create("protobuf", ProtobufConfigurator, project, fileResolver);
+
+    project.pluginManager.apply(ComponentModelBasePlugin)
+  }
+
+  static class Rules extends RuleSource {
```

Comment:
You got it mostly right. The rules are acting on the model. You can declare new element on the model with the @Model annotation. The @Default, @Mutate and @Finalize are simply to control when these run are applied on the model object. Each annotated rule method follow this simple convention: The first argument is the subject of the rule (what is been modified) and the following argument are the inputs of the rule. The argument for the @Model rule are all inputs (no subject as it's the return value). All input are guarantee to be completely configured before been used. Gradle take care of the dependencies between all the rule.

As for when, what and who of the rules, it's all handled by the internal of Gradle. When a plugin is applied, Gradle will look at the class and find the class extending the RuleSource. Then, it will scrap all the prototype of the annotated method and build the rules and it's dependencies. When Gradle need a specific element of the model, for example you want to execute a specific task, Gradle will call only the required rule (method) to configure that specific model.

This effectively make Gradle extremely lazy at the runtime level because it will only invoke what needs to be configure. This mainly the evaluation phase but does help with the execution phase. It prevent the old problem of "configuring the world" just to execute a simple task that is fairly independent.

The following videos really helped me to have a deep understanding of the model:
 - https://gradle.org/the-new-gradle-model/
 - https://gradle.org/rule-based-model-configuration-workshop-part-1/
 - https://gradle.org/rule-based-model-configuration-workshop-part-2/

I hope this help, don't hesitate if you have other question. I prefer you clearly understand.